File: C++/graph/14888.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

int n;
vector<int> v;
int arr[4];	// + - * /
int ans_min = 1000000001;
int ans_max = -1000000001;

void dfs(int idx, int result, int a1, int a2, int a3, int a4) {

	if (idx == n) {
		ans_min = min(ans_min, result);
		ans_max = max(ans_max, result);
		return;
	} 

	if (a1 > 0) {
		dfs(idx + 1, result + v[idx], a1 - 1, a2, a3, a4);
	}

	if (a2 > 0) {
		dfs(idx + 1, result - v[idx], a1, a2 - 1, a3, a4);
	}

	if (a3 > 0) {
		dfs(idx + 1, result * v[idx], a1, a2, a3 - 1, a4);
	}
	
	if (a4 > 0) {
		dfs(idx + 1, result / v[idx], a1, a2, a3, a4 - 1);
	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	
	cin >> n;

	for (int i = 0; i < n; i++) {
		int n2;
		cin >> n2;
		v.push_back(n2);
	}
	
	for (int i = 0; i < 4; i++) {
		cin >> arr[i];
	}

	dfs(1, v[0], arr[0], arr[1], arr[2], arr[3]);

	cout << ans_max << '\n' << ans_min << '\n';
	return 0;
}

File: Python/BFSDFS/14888.py
import sys
input = sys.stdin.readline

global ans_min, ans_max

ans_min = 10000000001
ans_max = -10000000001

n = int(input())
number_list = list(map(int, input().split()))
cal_info = list(map(int, input().split()))   # 덧셈, 뺄셈, 곱셈, 나눗셈


def dfs(idx, result, plus_cnt, minus_cnt, mul_cnt, div_cnt):
    global ans_max, ans_min

    if idx == n:
        ans_min = min(ans_min, result)
        ans_max = max(ans_max, result)
        return

    if plus_cnt > 0:
        dfs(idx+1, result + number_list[idx],
            plus_cnt-1, minus_cnt, mul_cnt, div_cnt)

    if minus_cnt > 0:
        dfs(idx+1, result - number_list[idx],
            plus_cnt, minus_cnt - 1, mul_cnt, div_cnt)

    if mul_cnt > 0:
        dfs(idx+1, result * number_list[idx],
            plus_cnt, minus_cnt, mul_cnt - 1, div_cnt)

    if div_cnt > 0:
        result_copy = result

        if result_copy < 0:
            result_copy *= -1
            result_copy = result_copy // number_list[idx]
            result_copy *= -1
        else:
            result_copy = result_copy // number_list[idx]

        dfs(idx+1, result_copy,
            plus_cnt, minus_cnt, mul_cnt, div_cnt - 1)


dfs(1, number_list[0], cal_info[0], cal_info[1], cal_info[2], cal_info[3])

print(ans_max)
print(ans_min)
